vsintegration/src/vs/FsPkgs/FSharp.Project/Common.Source.CSharp/Project/TypeProviderContainingReferenceNode.cs

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Open Technologies, Inc.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.VisualStudio.FSharp.ProjectSystem
+{
+    [CLSCompliant(false), ComVisible(true)]
+    public abstract class TypeProviderContainingReferenceNode : ReferenceNode
```

Comment:
shared is ContainsTypeProvider (2 lines, check -> redraw) and ImageIndex ( 2 lines ) and the field. The only real shared code is the const imageindex of icon.
It is not enough to require a new class, deeper hierarchy , more difficult read the code.